react/getting_started_with_react/passing_data_between_components/passing_data_between_components.md

```diff
@@ -0,0 +1,174 @@
+  )
+}
+
+export default function App() {
+  return (
+    <div>
+      <Button />
+      <Button2 />
+      <Button />
+    </div>
+  )
+}
+~~~
+This may not seem like a huge deal right now, but what if we had 10 buttons, each one having different text, fonts, colors, sizes, and any other variation you can think of. Creating a new component for each of these button variations would very quickly lead to a LOT of code duplication. 
+
+Let's see how by using props, we can account for any number of variations with a *single* button component.
+
+~~~jsx
+function Button(props) {
+
+  const buttonStyle = {
+    color: props.color,
+    fontSize: props.fontSize + 'px'
+  };
+
+  return (
+    <button style={buttonStyle}>{props.text}</button>
+  )
+}
+
+export default function App() {
+  return (
+    <div>
+      <Button text="Click Me!" color="blue" fontSize={12} />
+      <Button text="Don't Click Me!" color="red" fontSize={12} />
+      <Button text="Click Me!" color="blue" fontSize={20} />
+    </div>
+  );
+}
+~~~
+There are a few things going on here.
+
+- The `Button` functional component now receives `props` as a function argument. The individual properties are then referenced within the component via `props.propertyName`.
+- When rendering the `Button` components within `App`, the `prop` values are defined on each component.
+- Inline styles are dynamically generated and then applied to the `button` element.
+
+### Prop Destructuring
+
+A very common pattern you will come across in React is prop [destructuring](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment). Unpacking your props in the component arguments allows for more concise and readable code. Check out prop destructing in action in the example below.
+
+~~~jsx
+function Button({ text, color, fontSize }) {
+  const buttonStyle = {
+    color: color,
+    fontSize: fontSize + "px"
+  };
+
+  return <button style={buttonStyle}>{text}</button>;
+}
+
+export default function App() {
+  return (
+    <div>
+      <Button text="Click Me!" color="blue" fontSize={12} />
+      <Button text="Don't Click Me!" color="red" fontSize={12} />
+      <Button text="Click Me!" color="blue" fontSize={20} />
+    </div>
+  );
+}
+~~~
+### Default Props
+
+You may have noticed in the above examples that there is some repetition when defining props on the `Button` components within `App`. In order to stop repeating ourselves re-defining these common values, and to protect our application from undefined values, we can define default props that will be used by the component in the absence of supplied values.
+
+~~~jsx
+function Button({ text, color, fontSize }) {
+  const buttonStyle = {
+    color: color,
+    fontSize: fontSize + "px"
+  };
+
+  return <button style={buttonStyle}>{text}</button>;
+}
+
+Button.defaultProps = {
+  text: "Click Me!",
+  color: "blue",
+  fontSize: 12
+};
+
+export default function App() {
+  return (
+    <div>
```

Comment:
How would you feel about adding in this section that default props can be set in the component definition itself?

```jsx
function Button({ text, color, fontSize = 12 }) {
...
```

I'll let you decide if we want to add this information or not.

Reply:
TIL. I've added a brief mention.